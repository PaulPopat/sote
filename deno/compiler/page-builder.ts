import { PagesModel, TpeFile } from "../types/app.ts";
import { Iterate } from "../util/array.ts";
import { ParseTpeFile } from "./tpe-file-parser.ts";
import { NotUndefined } from "../util/object.ts";
import { StdComponents } from "../std-components.ts";
import { GetUsed } from "./tpe-component-applier.ts";

type TpeFileModel = {
  path: string;
  text: string;
  local_path: string;
};

export async function CompileApp(
  pages_files: AsyncGenerator<TpeFileModel>,
  components_files: AsyncGenerator<TpeFileModel>
): Promise<PagesModel> {
  const user_components = (
    await Iterate(components_files)
      .Select(async (f) => {
        try {
          return {
            data: await ParseTpeFile(f.text, f.local_path),
            url: f.path,
          };
        } catch (err) {
          console.log("Failed to parse component " + f.path + " error below:");
          console.error(err);

          return undefined;
        }
      })
      .WhereIs(NotUndefined)
      .ToArray()
  ).reduce(
    (c, n) => ({
      ...c,
      [n.url.replace("/", "").replace(/[\/\\]/gm, "::")]: n.data,
    }),
    {} as Record<string, TpeFile>
  );

  for (const component in user_components) {
    console.log("Adding component " + component + " to the component list.");
  }

  const components = { ...StdComponents, ...user_components };
  let css_bundle = "";
  let js_bundle = "";
  let included = [] as string[];
  const bundler: Record<string, (() => void)[]> = {};
  const full = await Iterate(pages_files)
    .Select(async (f) => {
      try {
        return {
          model: await ParseTpeFile(f.text, f.local_path),
          url: f.path,
        };
      } catch (err) {
        console.log("Failed to parse page " + f.path + " error below:");
        console.error(err);

        return undefined;
      }
    })
    .WhereIs(NotUndefined)
    .Select(({ model, url }) => ({
      url,
      model: {
        ...model,
        used: GetUsed(model.xml_template, components),
      },
    }))
    .ToArray();
  const pages = await Iterate(full)
    .Select(async ({ url, model }) => {
      let add = [] as TpeFile[];
      for (const include of model.used) {
        if (included.find((i) => i === include)) {
          continue;
        }

        const component = components[include];
        if (!component) {
          throw new Error("Cannot find component");
        }

        const total = full.filter(
          (f) => f.model.used.find((c) => c === include) != null
        ).length;
        if (total > full.length * 0.8) {
          included = [...included, include];
          if (component.css?.trim() && component.css.trim() !== "undefined") {
            css_bundle += "\n" + component.css;
          }

          if (component.client_js?.trim()) {
            js_bundle += "\n" + component.client_js;
          }
        } else {
          add = [...add, component];
        }
      }

      return {
        url,
        model: {
          ...model,
          xml_template: model.xml_template,
          client_js: add.reduce(
            (c, n) => (c ?? "") + (n.client_js ?? ""),
            model.client_js ?? ""
          ),
          css: add.reduce((c, n) => (c ?? "") + (n.css ?? ""), model.css ?? ""),
          used: undefined,
        },
      };
    })
    .Select((page) => {
      if (!page.model.title) {
        throw new Error(`Page ${page.url} does not have a title`);
      }

      if (!page.model.description) {
        throw new Error(`Page ${page.url} does not have a description`);
      }

      if (Object.keys(page.model.server_js).find((s) => s !== "get")) {
        console.log(
          `There are redirected handlers for ${page.url}. These handlers use cookies to perform the GET redirect (stopping resubmission on page refresh). Make sure you have a valid cookie notice on your site for these handlers.`
        );
      }

      return {
        ...page,
        model: {
          ...page.model,
          server_js: {
            ...page.model.server_js,
            get: page.model.server_js.get || "return query",
          },
          client_js: page.model.client_js,
          css: page.model.css,
          title: page.model.title ?? "",
          description: page.model.description ?? "",
        },
      };
    })
    .ToArray();

  return {
    pages: await Promise.all(
      pages.map(async (p) => ({
        ...p,
        model: { ...p.model, css: p.model.css },
      }))
    ),
    css_bundle,
    js_bundle,
    components,
  };
}
