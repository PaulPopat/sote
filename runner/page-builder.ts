import { ToXml, XmlEscape } from "../compiler/xml-parser.ts";
import { PagesModel, TpeFile } from "../types/app.ts";
import { Options } from "../types/config.ts";
import { RemoveUrlParameters } from "../util/url.ts";
import { BuildTpe, ReduceText } from "./tpe-builder.ts";

export async function BuildPage(
  page: PagesModel["pages"][number],
  components: Record<string, TpeFile>,
  bundle_js: string,
  bundle_css: string,
  props: any,
  context: any,
  options: Options,
  can_include_ga: boolean
) {
  const reduce = (text: string) =>
    ReduceText(text, props, [{ name: "context", value: context }]);
  return [
    `<!DOCTYPE html>`,
    `<html${
      options.lang
        ? ` lang="${
            page.model.language
              ? XmlEscape(await reduce(page.model.language))
              : XmlEscape(options.lang)
          }"`
        : ""
    }${
      options.email
        ? ` xmlns:v="urn:schemas-microsoft-com:vml" xmlns:o="urn:schemas-microsoft-com:office:office"`
        : ``
    }>`,
    `<head>`,
    `<meta charset="utf-8" />`,
    ...(options.google_tracking_id
      ? [
          `<script async src="https://www.googletagmanager.com/gtag/js?id=${options.google_tracking_id}"></script>`,
          !can_include_ga
            ? `<script>window.GAEnabled=false;function gtag(){dataLayer.push(arguments);}function EnableGA(){window.dataLayer=window.dataLayer||[];gtag('js',new Date());gtag('config','${options.google_tracking_id}');var CookieDate=new Date;CookieDate.setFullYear(CookieDate.getFullYear()+10);document.cookie="track-ga=true;expires="+CookieDate.toUTCString()+";";window.GAEnabled=true;}</script>`
            : `<script>window.dataLayer=window.dataLayer||[];gtag('js',new Date());gtag('config','${options.google_tracking_id}');window.GAEnabled=true;</script>`,
        ]
      : []),
    `<title>${XmlEscape(await reduce(page.model.title))}</title>`,
    `<meta name="description" content="${XmlEscape(
      await reduce(page.model.description)
    )}"/>`,
    options.author ? `<meta name="author" content="${options.author}"/>` : "",
    ...(options.email
      ? [
          `<!--[if mso]><xml><o:OfficeDocumentSettings><o:PixelsPerInch>96</o:PixelsPerInch></o:OfficeDocumentSettings></xml><![endif]-->`,
          '<meta http-equiv="X-UA-Compatible" content="IE=edge"/>',
          '<meta name="viewport" content="width=device-width,initial-scale=1"/>',
          '<meta name="x-apple-disable-message-reformatting"/>',
          '<meta name="color-scheme" content="light dark"/>',
          '<meta name="supported-color-schemes" content="light dark"/>',
        ]
      : [
          `<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"/>`,
        ]),
    ...(options.favicon
      ? options.favicon.map(
          (f) => `<link rel="icon" href="${f.path}" sizes="${f.size}"/>`
        )
      : []),

    ...(options.behavior_in_tag
      ? [
          page.model.css
            ? `<style type="text/css">${page.model.css}</style>`
            : "",
          bundle_css ? `<style type="text/css">${bundle_css}</style>` : "",
        ]
      : [
          page.model.css
            ? `<link rel="stylesheet" type="text/css" href="/css${RemoveUrlParameters(
                page.url
              )}.css"></link>`
            : "",
          bundle_css
            ? `<link rel="stylesheet" type="text/css" href="/css/index.bundle.css"></link>`
            : "",
        ]),
    `</head>`,

    `<body style="margin:0;padding:0;">`,
    ToXml(await BuildTpe(page.model.xml_template, components, props, context)),
    `</body>`,
    ...(options.behavior_in_tag
      ? [
          bundle_js
            ? `<script type="text/javascript">${bundle_js}</script>`
            : "",
          page.model.client_js
            ? `<script type="text/javascript">${page.model.client_js}</script>`
            : "",
        ]
      : [
          bundle_js
            ? `<script type="text/javascript" src="/js/common.bundle.js"></script>`
            : "",
          page.model.client_js
            ? `<script type="text/javascript" src="/js${RemoveUrlParameters(
                page.url
              )}.js"></script>`
            : "",
        ]),
    `</html>`,
  ].join("");
}
