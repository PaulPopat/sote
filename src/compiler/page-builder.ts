import { ParseTpeFile, TpeFile } from "./tpe-file-parser";
import { ApplyComponents, AppliedXmlNode } from "./tpe-component-applier";
import UglifyJS from "uglify-js";
import UglifyCss from "uglifycss";
import { NotUndefined } from "../utils/object";
import { StdComponents } from "../std-components";

type PageModel = {
  server_js: NodeJS.Dict<string>;
  client_js: string;
  xml_template: AppliedXmlNode[];
  css: string;
  title: string;
  description: string;
};

export type PagesModel = {
  pages: { url: string; model: PageModel }[];
  css_bundle: string;
  js_bundle: string;
};

type TpeFileModel = {
  path: string;
  text: string;
};

function MinifyJs(js: string, url: string) {
  const result = UglifyJS.minify(js);
  if (result.error) {
    console.log(`Failed to minify JS for ${url}. See error below.`);
    console.error(result.error);
    return js;
  }

  return result.code;
}

function MinifyCss(css: string) {
  return UglifyCss.processString(css, { maxLineLen: 500 });
}

export function CompileApp(
  pages_files: TpeFileModel[],
  components_files: TpeFileModel[],
  production: boolean
): PagesModel {
  const user_components = components_files
    .map((f) => {
      try {
        return { data: ParseTpeFile(f.text), url: f.path };
      } catch (err) {
        console.log("Failed to parse component " + f.path + " error below:");
        console.error(err);

        return undefined;
      }
    })
    .filter(NotUndefined)
    .reduce(
      (c, n) => ({
        ...c,
        [n.url.replace("/", "").replace(/[\/\\]/gm, "::")]: n.data,
      }),
      {} as NodeJS.Dict<TpeFile>
    );

  for (const component in user_components) {
    console.log("Adding component " + component + " to the component list.");
  }

  const components = { ...StdComponents, ...user_components };
  let css_bundle = "";
  let js_bundle = "";
  let included = [] as string[];
  const pages = pages_files
    .map((f) => {
      try {
        return { model: ParseTpeFile(f.text), url: f.path };
      } catch (err) {
        console.log("Failed to parse page " + f.path + " error below:");
        console.error(err);

        return undefined;
      }
    })
    .filter(NotUndefined)
    .map(({ model, url }) => ({
      url,
      model: {
        ...model,
        xml_template: ApplyComponents(model.xml_template, components),
      },
    }))
    .map(({ url, model }, _, full) => {
      let add = [] as TpeFile[];
      for (const include of model.xml_template.components) {
        if (included.find((i) => i === include)) {
          continue;
        }

        const component = components[include];
        if (!component) {
          throw new Error("Cannot find component");
        }

        const total = full.filter((f) =>
          f.model.xml_template.components.find((c) => c === include)
        ).length;
        if (total > full.length * 0.8) {
          included = [...included, include];
          if (component.css?.trim() && component.css.trim() !== "undefined") {
            css_bundle += "\n" + component.css;
          }

          if (component.client_js?.trim()) {
            js_bundle += "\n" + component.client_js;
          }
        } else {
          add = [...add, component];
        }
      }

      return {
        url,
        model: {
          ...model,
          xml_template: model.xml_template.tpe,
          client_js: add.reduce(
            (c, n) => (c ?? "") + (n.client_js ?? ""),
            model.client_js ?? ""
          ),
          css: add.reduce((c, n) => (c ?? "") + (n.css ?? ""), model.css ?? ""),
        },
      };
    })
    .map((page) => {
      if (!page.model.title) {
        throw new Error(`Page ${page.url} does not have a title`);
      }

      if (!page.model.description) {
        throw new Error(`Page ${page.url} does not have a description`);
      }

      return {
        ...page,
        model: {
          ...page.model,
          server_js: {
            ...page.model.server_js,
            get: page.model.server_js.get || "return query",
          },
          client_js:
            production && page.model.client_js?.trim()
              ? MinifyJs(page.model.client_js, page.url)
              : page.model.client_js?.trim()
              ? page.model.client_js
              : "",
          css:
            production && page.model.css?.trim()
              ? MinifyCss(page.model.css)
              : page.model.css?.trim()
              ? page.model.css
              : "",
          title: page.model.title ?? "",
          description: page.model.description ?? "",
        },
      };
    });

  return {
    pages: pages,
    css_bundle:
      production && css_bundle.trim() !== "undefined"
        ? MinifyCss(css_bundle)
        : css_bundle.trim() === "undefined"
        ? ""
        : css_bundle,
    js_bundle:
      production && js_bundle.trim() !== "undefined"
        ? MinifyJs(js_bundle, "Bundle")
        : js_bundle.trim() === "undefined"
        ? ""
        : js_bundle,
  };
}
