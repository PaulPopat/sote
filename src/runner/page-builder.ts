import { PagesModel } from "../compiler/page-builder";
import xmlescape from "xml-escape";
import { BuildTpe } from "./tpe-builder";
import { ToXml } from "../compiler/xml-parser";
import { Options } from "../file-system";
import { RemoveUrlParameters } from "../utils/url";

export async function BuildPage(
  page: PagesModel["pages"][number],
  bundle_js: string,
  bundle_css: string,
  props: any,
  context: any,
  options: Options
) {
  return [
    `<!DOCTYPE html>`,
    `<html${options.lang ? ` lang="${xmlescape(options.lang)}"` : ""}${
      options.email
        ? ` xmlns:v="urn:schemas-microsoft-com:vml" xmlns:o="urn:schemas-microsoft-com:office:office"`
        : ``
    }>`,
    `<head>`,
    `<meta charset="utf-8" />`,
    `<title>${xmlescape(page.model.title)}</title>`,
    `<meta name="description" content="${xmlescape(page.model.description)}"/>`,
    options.author ? `<meta name="author" content="${options.author}"/>` : "",
    ...(options.email
      ? [
          `<!--[if mso]><xml><o:OfficeDocumentSettings><o:PixelsPerInch>96</o:PixelsPerInch></o:OfficeDocumentSettings></xml><![endif]-->`,
          '<meta http-equiv="X-UA-Compatible" content="IE=edge"/>',
          '<meta name="viewport" content="width=device-width,initial-scale=1"/>',
          '<meta name="x-apple-disable-message-reformatting"/>',
          '<meta name="color-scheme" content="light dark"/>',
          '<meta name="supported-color-schemes" content="light dark"/>',
        ]
      : [
          `<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"/>`,
        ]),
    ...(options.favicon
      ? options.favicon.map(
          (f) => `<link rel="icon" href="${f.path}" sizes="${f.size}"/>`
        )
      : []),

    ...(options.behavior_in_tag
      ? [
          page.model.css
            ? `<style type="text/css">${page.model.css}</style>`
            : "",
          bundle_css ? `<style type="text/css">${bundle_css}</style>` : "",
        ]
      : [
          page.model.css
            ? `<link rel="stylesheet" type="text/css" href="/css${RemoveUrlParameters(
                page.url
              )}.css"></link>`
            : "",
          bundle_css
            ? `<link rel="stylesheet" type="text/css" href="/css/index.bundle.css"></link>`
            : "",
        ]),
    `</head>`,

    `<body style="margin:0;padding:0;">`,
    ToXml(
      await BuildTpe(page.model.xml_template, page.model.tree, props, context)
    ),
    `</body>`,
    ...(options.behavior_in_tag
      ? [
          bundle_js
            ? `<script type="text/javascript">${bundle_js}</script>`
            : "",
          page.model.client_js
            ? `<script type="text/javascript">${page.model.client_js}</script>`
            : "",
        ]
      : [
          bundle_js
            ? `<script type="text/javascript" src="/js/common.bundle.js"></script>`
            : "",
          page.model.client_js
            ? `<script type="text/javascript" src="/js${RemoveUrlParameters(
                page.url
              )}.js"></script>`
            : "",
        ]),
    `</html>`,
  ].join("");
}
