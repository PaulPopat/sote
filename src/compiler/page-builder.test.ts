import { CompileApp } from "./page-builder";

it("Compiles a basic page", () => {
  expect(
    CompileApp(
      [
        {
          path: "/test",
          text: `
<template>
  <div />
</template>
<title>A test page</title>
<description>A test description</description>`,
        },
      ],
      [],
      false
    )
  ).toMatchSnapshot();
});

it("Compiles client javascript", () => {
  expect(
    CompileApp(
      [
        {
          path: "/test",
          text: `
<template>
  <div />
</template>
<script area="client">
  console.log("Hello world")
</script>
<title>A test page</title>
<description>A test description</description>`,
        },
      ],
      [],
      false
    )
  ).toMatchSnapshot();
});

it("Minifies client js for production", () => {
  expect(
    CompileApp(
      [
        {
          path: "/test",
          text: `
<template>
  <div />
</template>
<script area="client">
  function thing() {
    console.log("hello world")
  }
</script>
<title>A test page</title>
<description>A test description</description>`,
        },
      ],
      [],
      true
    )
  ).toMatchSnapshot();
});

it("Compiles server get js", () => {
  expect(
    CompileApp(
      [
        {
          path: "/test",
          text: `
<template>
  <div />
</template>
<script area="server">
  console.log("Hello world");
  return {};
</script>
<title>A test page</title>
<description>A test description</description>`,
        },
      ],
      [],
      false
    )
  ).toMatchSnapshot();
});

it("Compiles method js for server", () => {
  expect(
    CompileApp(
      [
        {
          path: "/test",
          text: `
<template>
  <div />
</template>
<script area="server" method="post">
  console.log("Hello world");
  return {};
</script>
<title>A test page</title>
<description>A test description</description>`,
        },
      ],
      [],
      false
    )
  ).toMatchSnapshot();
});

it("Compiles page css", () => {
  expect(
    CompileApp(
      [
        {
          path: "/test",
          text: `
<template>
  <div />
</template>
<style>
  .test { display: block; }
</style>
<title>A test page</title>
<description>A test description</description>`,
        },
      ],
      [],
      false
    )
  ).toMatchSnapshot();
});

it("Bundles component css", () => {
  expect(
    CompileApp(
      [
        {
          path: "/test",
          text: `
<template>
  <test::component>
    <div />
  </test::component>
</template>
<style>
  .test { display: block; }
</style>
<title>A test page</title>
<description>A test description</description>`,
        },
      ],
      [
        {
          path: "/test/component",
          text: `
<template>
  <div>
    <children />
  </div>
</template>
<style>
  div { display: block; }
</style>
        `,
        },
      ],
      false
    )
  ).toMatchSnapshot();
});

it("Bundles component javascript", () => {
  expect(
    CompileApp(
      [
        {
          path: "/test",
          text: `
<template>
  <test::component>
    <div />
  </test::component>
</template>
<style>
  .test { display: block; }
</style>
<title>A test page</title>
<description>A test description</description>`,
        },
      ],
      [
        {
          path: "/test/component",
          text: `
<template>
  <div>
    <children />
  </div>
</template>
<script area="client">
  console.log("Hello world");
</script>
        `,
        },
      ],
      false
    )
  ).toMatchSnapshot();
});

it("Does not bundle css if the component is not always used", () => {
  expect(
    CompileApp(
      [
        {
          path: "/test",
          text: `
<template>
  <test::component>
    <div />
  </test::component>
</template>
<style>
  .test { display: block; }
</style>
<title>A test page</title>
<description>A test description</description>`,
        },
        {
          path: "/test",
          text: `
<template>
  <div />
</template>
<style>
  .test { display: block; }
</style>
<title>A test page</title>
<description>A test description</description>`,
        },
      ],
      [
        {
          path: "/test/component",
          text: `
<template>
  <div>
    <children />
  </div>
</template>
<style>
  div { display: block; }
</style>
        `,
        },
      ],
      false
    )
  ).toMatchSnapshot();
});

it("Bundles component javascript", () => {
  expect(
    CompileApp(
      [
        {
          path: "/test",
          text: `
<template>
  <test::component>
    <div />
  </test::component>
</template>
<style>
  .test { display: block; }
</style>
<title>A test page</title>
<description>A test description</description>`,
        },
        {
          path: "/test",
          text: `
<template>
  <div />
</template>
<style>
  .test { display: block; }
</style>
<title>A test page</title>
<description>A test description</description>`,
        },
      ],
      [
        {
          path: "/test/component",
          text: `
<template>
  <div>
    <children />
  </div>
</template>
<script area="client">
  console.log("Hello world");
</script>
        `,
        },
      ],
      false
    )
  ).toMatchSnapshot();
});