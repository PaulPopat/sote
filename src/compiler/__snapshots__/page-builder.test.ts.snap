// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bundles component css 1`] = `
Object {
  "css_bundle": "
div[data-specifier=\\"3a003c8ed08e0f1e53bff9cac752c55e\\"]{display:block;}",
  "js_bundle": "",
  "pages": Array [
    Object {
      "model": Object {
        "client_js": "",
        "css": ".test[data-specifier=\\"edf762dd4e455c036183858efa983eaf\\"]{display:block;}",
        "description": "A test description",
        "server_js": Object {
          "get": "return query",
        },
        "title": "A test page",
        "xml_template": Array [
          Object {
            "attributes": Object {
              "data-specifier": "3a003c8ed08e0f1e53bff9cac752c55e",
            },
            "children": Array [
              Object {
                "attributes": Object {
                  "data-specifier": "edf762dd4e455c036183858efa983eaf",
                },
                "children": Array [],
                "tag": "div",
              },
            ],
            "tag": "div",
          },
        ],
      },
      "url": "/test",
    },
  ],
}
`;

exports[`Bundles component javascript 1`] = `
Object {
  "css_bundle": "",
  "js_bundle": "
console.log(\\"Hello world\\");",
  "pages": Array [
    Object {
      "model": Object {
        "client_js": "",
        "css": ".test[data-specifier=\\"edf762dd4e455c036183858efa983eaf\\"]{display:block;}",
        "description": "A test description",
        "server_js": Object {
          "get": "return query",
        },
        "title": "A test page",
        "xml_template": Array [
          Object {
            "attributes": Object {},
            "children": Array [
              Object {
                "attributes": Object {
                  "data-specifier": "edf762dd4e455c036183858efa983eaf",
                },
                "children": Array [],
                "tag": "div",
              },
            ],
            "tag": "div",
          },
        ],
      },
      "url": "/test",
    },
  ],
}
`;

exports[`Bundles component javascript 2`] = `
Object {
  "css_bundle": "",
  "js_bundle": "",
  "pages": Array [
    Object {
      "model": Object {
        "client_js": "undefinedconsole.log(\\"Hello world\\");",
        "css": ".test[data-specifier=\\"edf762dd4e455c036183858efa983eaf\\"]{display:block;}",
        "description": "A test description",
        "server_js": Object {
          "get": "return query",
        },
        "title": "A test page",
        "xml_template": Array [
          Object {
            "attributes": Object {},
            "children": Array [
              Object {
                "attributes": Object {
                  "data-specifier": "edf762dd4e455c036183858efa983eaf",
                },
                "children": Array [],
                "tag": "div",
              },
            ],
            "tag": "div",
          },
        ],
      },
      "url": "/test",
    },
    Object {
      "model": Object {
        "client_js": "",
        "css": ".test[data-specifier=\\"edf762dd4e455c036183858efa983eaf\\"]{display:block;}",
        "description": "A test description",
        "server_js": Object {
          "get": "return query",
        },
        "title": "A test page",
        "xml_template": Array [
          Object {
            "attributes": Object {
              "data-specifier": "edf762dd4e455c036183858efa983eaf",
            },
            "children": Array [],
            "tag": "div",
          },
        ],
      },
      "url": "/test",
    },
  ],
}
`;

exports[`Compiles a basic page 1`] = `
Object {
  "css_bundle": "",
  "js_bundle": "",
  "pages": Array [
    Object {
      "model": Object {
        "client_js": "",
        "css": "",
        "description": "A test description",
        "server_js": Object {
          "get": "return query",
        },
        "title": "A test page",
        "xml_template": Array [
          Object {
            "attributes": Object {},
            "children": Array [],
            "tag": "div",
          },
        ],
      },
      "url": "/test",
    },
  ],
}
`;

exports[`Compiles client javascript 1`] = `
Object {
  "css_bundle": "",
  "js_bundle": "",
  "pages": Array [
    Object {
      "model": Object {
        "client_js": "console.log(\\"Hello world\\")",
        "css": "",
        "description": "A test description",
        "server_js": Object {
          "get": "return query",
        },
        "title": "A test page",
        "xml_template": Array [
          Object {
            "attributes": Object {},
            "children": Array [],
            "tag": "div",
          },
        ],
      },
      "url": "/test",
    },
  ],
}
`;

exports[`Compiles method js for server 1`] = `
Object {
  "css_bundle": "",
  "js_bundle": "",
  "pages": Array [
    Object {
      "model": Object {
        "client_js": "",
        "css": "",
        "description": "A test description",
        "server_js": Object {
          "get": "return query",
          "post": "console.log(\\"Hello world\\");
  return {};",
        },
        "title": "A test page",
        "xml_template": Array [
          Object {
            "attributes": Object {},
            "children": Array [],
            "tag": "div",
          },
        ],
      },
      "url": "/test",
    },
  ],
}
`;

exports[`Compiles page css 1`] = `
Object {
  "css_bundle": "",
  "js_bundle": "",
  "pages": Array [
    Object {
      "model": Object {
        "client_js": "",
        "css": ".test[data-specifier=\\"edf762dd4e455c036183858efa983eaf\\"]{display:block;}",
        "description": "A test description",
        "server_js": Object {
          "get": "return query",
        },
        "title": "A test page",
        "xml_template": Array [
          Object {
            "attributes": Object {
              "data-specifier": "edf762dd4e455c036183858efa983eaf",
            },
            "children": Array [],
            "tag": "div",
          },
        ],
      },
      "url": "/test",
    },
  ],
}
`;

exports[`Compiles server get js 1`] = `
Object {
  "css_bundle": "",
  "js_bundle": "",
  "pages": Array [
    Object {
      "model": Object {
        "client_js": "",
        "css": "",
        "description": "A test description",
        "server_js": Object {
          "get": "console.log(\\"Hello world\\");
  return {};",
        },
        "title": "A test page",
        "xml_template": Array [
          Object {
            "attributes": Object {},
            "children": Array [],
            "tag": "div",
          },
        ],
      },
      "url": "/test",
    },
  ],
}
`;

exports[`Does not bundle css if the component is not always used 1`] = `
Object {
  "css_bundle": "",
  "js_bundle": "",
  "pages": Array [
    Object {
      "model": Object {
        "client_js": "",
        "css": ".test[data-specifier=\\"edf762dd4e455c036183858efa983eaf\\"]{display:block;}div[data-specifier=\\"3a003c8ed08e0f1e53bff9cac752c55e\\"]{display:block;}",
        "description": "A test description",
        "server_js": Object {
          "get": "return query",
        },
        "title": "A test page",
        "xml_template": Array [
          Object {
            "attributes": Object {
              "data-specifier": "3a003c8ed08e0f1e53bff9cac752c55e",
            },
            "children": Array [
              Object {
                "attributes": Object {
                  "data-specifier": "edf762dd4e455c036183858efa983eaf",
                },
                "children": Array [],
                "tag": "div",
              },
            ],
            "tag": "div",
          },
        ],
      },
      "url": "/test",
    },
    Object {
      "model": Object {
        "client_js": "",
        "css": ".test[data-specifier=\\"edf762dd4e455c036183858efa983eaf\\"]{display:block;}",
        "description": "A test description",
        "server_js": Object {
          "get": "return query",
        },
        "title": "A test page",
        "xml_template": Array [
          Object {
            "attributes": Object {
              "data-specifier": "edf762dd4e455c036183858efa983eaf",
            },
            "children": Array [],
            "tag": "div",
          },
        ],
      },
      "url": "/test",
    },
  ],
}
`;

exports[`Minifies client js for production 1`] = `
Object {
  "css_bundle": "",
  "js_bundle": "",
  "pages": Array [
    Object {
      "model": Object {
        "client_js": "function thing(){console.log(\\"hello world\\")}",
        "css": "",
        "description": "A test description",
        "server_js": Object {
          "get": "return query",
        },
        "title": "A test page",
        "xml_template": Array [
          Object {
            "attributes": Object {},
            "children": Array [],
            "tag": "div",
          },
        ],
      },
      "url": "/test",
    },
  ],
}
`;
